### DISCUSSION TITLE
Is StreamDeck+ supported?

### DISCUSSION THREAD

**Discussion by LogicalUnit** [streamdeck-linux-gui/streamdeck-linux-gui#130] [Ideas / Feature Requests]
Newer model StreamDeck+ with dials as well as buttons. On my system I've got the UI running but terminal says No Stream Deck(s) detected. Attach a Stream Deck.

https://www.elgato.com/us/en/p/stream-deck-plus-black

**Comment by coolapso** [comment]
No its not supported at the moment, however it seems to be already supported on the upstream library.  just think that maybe no one in the usual contributors has one to work with and test =/ 

Guess it will depend on someone with one having the knowledge and will to contribute and add the support. Unfortunately I lack both, but mostly lack the device as the knowledge can be acquired :D

**Reply by BiosElement** [reply]
If you could point me in the direction of wherever device matching and support is handled, I would be happy to see if I could help as I do own a StreamDeck Plus and full-time linux is limited by it currently. -_-

**Comment by slovia** [comment]
I am very interested in this too. Is there anything I can do to help? Support for Stream Deck+ would be awesome!

**Comment by coolapso** [comment]
For those asking "What can i do?" in this case the answer is everything ... The upstream library here: https://github.com/abcminiuser/python-elgato-streamdeck already supports it so guess that's the starting point. 

@BiosElement  
> If you could point me in the direction of wherever device matching and support is handled, I would be happy to see if I could help as I do own a StreamDeck Plus and full-time linux is limited by it currently. -_-

here I'd argue if you can't figure it out then you probably don't have the knowledge for doing it.

**Comment by ajbucci** [comment]
The dev of the upstream library has pinned a new release that has support for the plus. We will need to bump the dependency and test to make sure there are no regressions.

Currently handling of supported device detection is handled within the upstream library.

It's possible that just bumping the dependency version will allow the non-rotary buttons to work immediately, but I'll need to test when I get home from the holidays. I will begin work on integrating full support starting next week.

**Reply by slovia** [reply]
This is awesome news! Thank you!
Have wonderful remaining holidays!
PS: I am new to Linux and I have no coding skills so I can't help with anything and even "bumping the dependency version" is nothing I have the knowledge to do. But if I can donate a buck or two I am very willing to help that way! Thanks!

**Reply by vihu** [reply]
> It's possible that just bumping the dependency version will allow the non-rotary buttons to work immediately, but I'll need to test when I get home from the holidays. I will begin work on integrating full support starting next week.

Can confirm this is indeed true, I simply updated the pinned dep like so:

```diff
diff --git a/pyproject.toml b/pyproject.toml
index 3692335..6071f97 100644
--- a/pyproject.toml
+++ b/pyproject.toml
@@ -13,7 +13,7 @@ packages = [
 
 [tool.poetry.dependencies]
 python = ">=3.8,<3.12"
-streamdeck = "^0.9.3"
+streamdeck = "^0.9.5"
 pillow = "^9.4.0"
 pyside6 = "^6.4.2"
 CairoSVG = "^2.5.2"
```

And I have working buttons (no rotary dials obviously):

![image](https://github.com/streamdeck-linux-gui/streamdeck-linux-gui/assets/3199183/b8c2dbdf-d322-4ef4-9b6a-d5019ef97815)

**Reply by LogicalUnit** [reply]
Awesome, this is great news. I had a go at building from source with the updated dependency, but it still wouldn't detect my StreamDeck+ device. It would be great if we could merge this functionality into the project as a temporary measure.

**Reply by vihu** [reply]
I had to double check the udev rules, the docs are slightly off I think. This is what I put in my `/etc/udev/rules/70-streamdeck.rules` for reference:

```
SUBSYSTEM=="usb", ATTRS{idVendor}=="0fd9", TAG+="uaccess"
KERNEL=="uinput", SUBSYSTEM=="input", TAG+="uaccess"
```

Then do `sudo udevadm trigger` and reconnect your SD+. That might just do the trick.

**Reply by LogicalUnit** [reply]
![Screenshot from 2023-12-31 14-39-51](https://github.com/streamdeck-linux-gui/streamdeck-linux-gui/assets/6601496/8eb1a956-452f-4ac5-b882-f82b2adb1dab)


Yeah! I got it working! I updated the udev rules, but was still having trouble. I think the $HOME variable in the python install step was putting the bins/libs in the wrong place, and I was re-running the existing streamdeck/streamdeckc instead of the ones I had just compiled. Now I've learned about the which command lol. Linux is awesome, thank you!

**Reply by ajbucci** [reply]
To be fair to Windows, it has "where" instead ;)

**Reply by ajbucci** [reply]
Glad to hear that bumping the dependency itself gets the regular buttons working on the plus! 

If you all wouldn't mind testing everything out over the next few days to see that all the functionality works as intended with the dependency version change I'd appreciate it! _Especially_ if you all have non-plus versions of the deck you can test with. I don't currently have access to either of my decks to test.

**Comment by LogicalUnit** [comment]
Okay, I've had a chance to play with it for a bit. The buttons can certainly be configured as expected: Images, Backgrounds, Labels, Pages, States, and Commands all work.

However, the Press Keys and Write Text commands both fail with a similar error

![Screenshot from 2024-01-02 14-14-04](https://github.com/streamdeck-linux-gui/streamdeck-linux-gui/assets/6601496/0f1846de-8e97-474d-b6f3-bfd93e0a5ca6)

**Reply by ajbucci** [reply]
> Okay, I've had a chance to play with it for a bit. The buttons can certainly be configured as expected: Images, Backgrounds, Labels, Pages, States, and Commands all work.
> 

Thanks for checking and glad to hear things are working.

> However, the Press Keys and Write Text commands both fail with a similar error
> 

I believe those errors are unrelated. They're due to a recent PR trying to get those features working more consistently on Wayland.

**Reply by ajbucci** [reply]
@LogicalUnit here is a quick fix for that error https://github.com/streamdeck-linux-gui/streamdeck-linux-gui/pull/143

**Reply by LogicalUnit** [reply]
@ajbucci I can confirm that both the Press Keys and Write Text functionality have been fixed. What a legend! :)

**Comment by 03l54rd1n3** [comment]
Are there any progress updates on getting the dials or touchscreen working?

**Reply by ajbucci** [reply]
I have the GUI set up for the buttons only, but that was the easy part. Still need to sort out the data model / right-side pane GUI. Then I'll need to hook everything in to the existing code without breaking anything (this is the hardest part)... I have some of this already finished but perhaps just 25-30%... I haven't had too much time to work on it lately. I may put what I have so far up in a WIP PR.

**Reply by cking** [reply]
that would be highly appreciated from me <3

**Reply by mboelen** [reply]
The underlying [library](https://github.com/abcminiuser/python-elgato-streamdeck) (0.95) does support the Stream Deck + now correctly. I just tested it and it seems to support the dials and touch screen as well. Actually thought the screen was just a display, but looking at the code I saw it's more than that. So it looks like it is now "only" a matter of adding this support to this project.

I have Python knowledge, but not familiar with UI design. So at this moment I can't contribute (yet) otherwise I would have done some fiddling and testing.

**Comment by Giraut** [comment]
I have a Stream Deck Plus and I'm planning to add support for the touchscreen and the dials soon. In fact, I have [another project](https://github.com/Giraut/freecad_streamdeck_addon) that uses the Stream Deck that already supports the dials in the Plus. It's not terribly complicated.

It's just that I need to think about how to use the touchscreen and the dials. The buttons are easy: one image, one button, click... It's quite straightforward. But there are a million ways to exploit a touchscreen and 4 clicky dials, and I'm not sure which is best. My own project is stalled with that for the same reason.

Elgato apparently has this plugin system whereby one loads a plugin that does something interesting with the device. There is no such thing in streamdeck_ui / streamdeck_linux_ui, and I'm not willing to embark on coding a plugin infrastructure for it. But I will implement something useful that uses the dials, that I would want to use myself. Probably some generic command spawning with an expanded variable to pass the number of dial clicks or something.

As for the touchscreen, I'll probably make it emulate another 4 to 8 buttons, with swiping to change pages.

**Reply by ajbucci** [reply]
I've already embarked on this though I haven't had time to finish. I've run into the same issue... Issuing the commands is easy, but deciding how to support the interaction between dial/touchscreen (as well as what functionality to expose to the dial rotation) has been a speed bump. My gut is to just partition the touch screen into 4  columns (one for each dial) and use the dial settings to set the touchscreen image per column for now. Perhaps allowing swipes/taps of the touchscreen to change pages. 

These are all linked:
- handlers for rotate cw/ccw (this should be straightforward)
- how should the GUI be organized to reduce clutter and allow customization of the 2 additional available handlers per button/action? What actions do we allow the dial rotation to perform? All the currently supported actions or a subset? (I'm getting hung up on this)
- any new actions need their configurations saved from the GUI to the settings json and loaded to the GUI from the settings json. Models will be needed for the DialState and TouchscreenState
- the configuration pane itself needs to be linked with the model / settings etc. and there is a lot of plumbing that needs to be added/modified in order to get this working on the back end. Currently the code that loads the configuration panes assumes there is only one type

The buttons themselves:
- dial / touchscreen layout for the GUI (I've already completed this)
- dial / touchscreen button objects that support drag and drop etc in the GUI (I've also already completed this)

There's probably more that I can't remember or haven't considered yet, it has been a while since I've had a chance to look at this.

**Reply by LogicalUnit** [reply]
The feature I would like from the dials is the ability to control my Elgato Key Light Air. Currently, I can use the Elgato mobile app or this Linux project to control it: https://github.com/justinforlenza/keylight-control
The ability to adjust brightness and temperature with a dial would be doooope.

**Reply by Giraut** [reply]
> I've run into the same issue... 

This is what I plan to do personally (or at least, what I *was* planning, since you've already started: no point in doing the work twice 🙂):

* Touchscreen
  - Add 8 buttons with the touchscreen: the touchscreen is 800x100, so it lends itself naturally to eight 100x100 buttons.
  - The touchscreen buttons act as their own mini separate Stream Deck. That is, they're totally separate from from the 8 physical buttons on top. In streamdeck-ui / streamdeck-linux-gui, they'll have their own DisplayGrid and they'll function exactly like the 8 physical buttons. The only thing that requires is a compositor that recreates and updates the entire LCD image from the 100x100 individual logical buttons. That's a very minor change.
  - The touchscreen DisplayGrid has a new filter that flips the page when swiping left or right. Again, a very minor change.

* Dials
  - Each dial is associated with a command that gets called when the dial is acted upon - like the buttons - but instead of just simply calling the command, the command is passed 2 arguments: one for the up/down state, one for the number of turning clicks. For example:
    - Turn the dial slowly clockwise: `/path/command --press 0 --turn 1`
    - Turn the dial slowly counterclockwise: `/path/command --press 0 --turn -1`
    - Turn the dial quickly clockwise: `/path/command --press 0 --turn 12`
    - Press the dial: `/path/command --press 1 --turn 0`
    - Turn the dial while pressing the dial: `/path/command --press 1 --turn 1`

The touchscreen changes are trivial and require no addition to the GUI because the touchscreen is treated like another set of keys.

The dial changes essentially externalize the dial actions to the script. Meaning the streamdeck program doesn't have to decide what to do with the dials: the scripts do. And because it uses the regular command line argument mechanism, no plugin infrastructure is requires, and whoever wants to use them doesn't have to code a plugin.

The real inconvenience of the dial system - at least for me - is that it requires additions to the GUI to configure the commands to associate to each dial, and I hate Qt 🙂 It's really the only thing that holds me back. But I might do the touchscreen thing. If you've already done work and you have a branch, I could branch off of yours and do it there if you'd like.

**Reply by mboelen** [reply]
I was personally thinking to define one dial for switching between pages. Most likely I don't need 8 additional buttons as part of the touch screen, but it would be great to have this as an option in the GUI.

Would be great if the GUI is similar to the buttons, but where you can select an event as mentioned by @Giraut and the related action (firing a command, send text, switch page, change image on touchscreen). Similar to the events displayed by the [example file](https://github.com/abcminiuser/python-elgato-streamdeck/blob/master/src/example_plus.py).

If I can do anything to help testing or brainstorm, just let me know! Would love to see this device fully supported, so that I can recommend it to others.

**Reply by Giraut** [reply]
> I was personally thinking to define one dial for switching between pages

I can give some insight on that one: my own Stream Deck project does that with the Stream Deck + and it turns out to be rather inconvenient. I noticed it myself when I implemented it, and a couple users reported it as well. So much so that I'm considering using the swipe function of the touchscreen to swipe pages instead (because I don't want to sacrifice 2 regular buttons out of the mere 8 on that particular Deck to do that).

The problem basically is that the dials are too sensitive between clicks. You can't conveniently change pages accurately: you have to pay attention to how much you turn the dial exactly.

I thought it would be better, but it really isn't. You can see - and feel for yourself - what it looks like to changes pages with the dial in the video [here](https://www.reddit.com/r/FreeCAD/comments/1b7ql71/stream_deck_addon_now_has_basic_support_for_the/?utm_source=share&utm_medium=web3x&utm_name=web3xcss&utm_term=1&utm_content=share_button).

**Reply by coolapso** [reply]
Well ... just dropping my 2 cents .. .why not expose that to users and let them decide if: 

Display: 
4 buttons
8 Buttons
Swipe

As for the dials ... no clue .. but instead of  "Should it do X or should it do Y", just expose that and decouple the code in a way that its easy to add / remove / change functionality. 

Don't forget this application is used for more people, limiting to only one use case is limiting the application and other users. 

OK it might be too much to expose everything in one go? As I said before just make sure whatever you code allows for further expansion/ removal and don't assume that's the only use case in the whole app. 

other than that ... I am here just to make decisions and help with as much as I can but not actual code :)

**Reply by Giraut** [reply]
If you try to expose too much, it gets complicated. The more complicated it gets, the fewer people will want to use it.

The dials calling external scripts with arguments is as much as I'd personally want to burden the average user with. I don't think the touchscreen should be configuration beyond - like you said - 4 or 8 buttons, and possibly extending the main set of buttons or separate. The swipe feature should change pages and that's it. If you try to expose the touchscreen too much, it's going to become Rube Goldberg-ish.

Anyway, I reckon the easiest is to just get coding and see what users think later. I have a feeling it's the sort of thing that will become very annoying if you overthink it.

**Reply by coolapso** [reply]
Maybe I was not clear :D 
I don't want to expose to much ... and I agree. but expose too little and you leave users behind :) 

 The swipe can always work regardless if you have 4 or 8 Buttons right? 
 
 > Anyway, I reckon the easiest is to just get coding and see what users think later. I have a feeling it's the sort of thing that will become very annoying if you overthink it.
 
 100% Agree better do it incrementally and have little than try to do too much and end up having nothing :)

**Reply by Giraut** [reply]
> The swipe can always work regardless if you have 4 or 8 Buttons right?

Actually the touchscreen is just that - a 800x100 touchscreen. There are no real buttons on it, just a 800x100 image. The buttons I was talking about were logical buttons composited and displayed on the screen, because I figured I'd use the screen for buttons. But technically, if you wanted to play a video on it, you could. And of course, the swipe feature is just a touchscreen swipe - i.e. it reports where the swipe starts and where it ends on the screen.

**Reply by coolapso** [reply]
ok cool! then i'd say the swipe left or right to change page is always "enabled" and users can choose if they want 4 or 8 buttons or maybe just a single button with whatever they want on the screen? :shrug: leave it up to you. but this is pretty much what I meant. 

for the dials ... I am not entirely sure, but start it as simple as you can, more worried that the code is flexible enough to be maintainable and anyone can keep building on top of it. 

if it helps, If I had one of those streamdeck+  my use case would probably be mostly "stateless" operations:

* Change volume (rotary) / mute / unmute press
* Maybe make api calls to home assistant to dimm lights (rotary) / presses to turn on or off
* zoom in / zoom out the terminal when doing screen sharing 

not entirely sure what else.

**Reply by LogicalUnit** [reply]
Any development on this? I've built version 4.1.2, but the touchscreen and dials are not working yet.

**Reply by Giraut** [reply]
Hey hi,

Yeah sorry but I'm not gonna do it. The code is a needlessly complicated pile, Qt is kind of a horrible toolkit to work with anyway, adding touchscreen and dial support cleanly requires fairly involved additions to the UI and I don't feel that this codebase is worth working on any further. The alternative is a total rewrite and I'm not interested in doing that either. So I gave up on it. Not worth my time.

**Reply by coolapso** [reply]
https://github.com/streamdeck-linux-gui/streamdeck-linux-gui/discussions/203
